route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -735,4 +749,56 @@ void setManeuverLaneletIds(cav_msgs::Maneuver& mvr, lanelet::Id start_id, lanele
         tf2_listener_.reset(new tf2_ros::TransformListener(tf2_buffer_));
         tf2_buffer_.setUsingDedicatedThread(true);
     }
+
+    void RouteFollowingPlugin::returnToShortestPath(const lanelet::ConstLanelet &current_lanelet)
+    {
+        auto original_shortestpath = wm_->getRoute()->shortestPath();
+        ROS_DEBUG_STREAM("The vehicle has left the shortest path");
+        auto routing_graph = wm_->getMapRoutingGraph();
+
+        // In order to return to the shortest path, the closest future lanelet on the shortest path needs to be found.
+        // That is the following lanelet of the adjacent lanelet of the current lanelet.
+        auto adjacent_lanelets = routing_graph->besides(current_lanelet);
+        if (!adjacent_lanelets.empty())
+        {
+            for (auto adjacent:adjacent_lanelets)
+            {
+                if (shortest_path_set_.find(adjacent.id())!=shortest_path_set_.end())
+                {
+                    auto following_lanelets = routing_graph->following(adjacent);
+                    auto target_following_lanelet = following_lanelets[0];
+                    ROS_DEBUG_STREAM("The target_following_lanelet id is: " << target_following_lanelet.id());
+                    lanelet::ConstLanelets interm;
+                    interm.push_back(static_cast<lanelet::ConstLanelet>(target_following_lanelet));
+                    // a new shortest path, via the target_following_lanelet is calculated and used an alternative shortest path
+                    auto new_shortestpath = routing_graph->shortestPathVia(current_lanelet, interm, original_shortestpath.back());
+                    ROS_DEBUG_STREAM("a new shortestpath is generated to return to original shortestpath");
+                    // routeCb is called to update latest_maneuver_plan_
+                    if (new_shortestpath) this->latest_maneuver_plan_ = routeCb(new_shortestpath.get());
+                    break;
+                }
+                else
+                {
+                    // a new shortest path, via the current_lanelet is calculated and used an alternative shortest path
+                    lanelet::ConstLanelets new_interm;
+                    new_interm.push_back(static_cast<lanelet::ConstLanelet>(current_lanelet));
+                    auto new_shortestpath = routing_graph->shortestPathVia(current_lanelet, new_interm, original_shortestpath.back());
+                    ROS_DEBUG_STREAM("Cannot return to the original shortestpath from adjacent lanes, so a new shortestpath is generated");
+                    // routeCb is called to update latest_maneuver_plan_
+                    if (new_shortestpath) this->latest_maneuver_plan_ = routeCb(new_shortestpath.get());
+                }
+            }
+            
+        }
+        else
+        {
+            ROS_WARN_STREAM("Alternative shortest path cannot be generated");
```

Comment:
Should this throw an exception as well? If not, how would the plugin behave if we don't update the shortest path?

Reply:
there is a pose callback in route node, that if the vehicles is in an un-routable position will throw an exception. I suggests we let that handle this situation.